sql/src/main/java/org/elasticsearch/action/bulk/BulkRetryCoordinator.java

```diff
@@ -21,132 +21,160 @@
 
 package org.elasticsearch.action.bulk;
 
+import io.crate.action.LimitedExponentialBackoff;
+import io.crate.exceptions.Exceptions;
 import io.crate.executor.transport.ShardRequest;
 import io.crate.executor.transport.ShardResponse;
 import org.elasticsearch.action.ActionListener;
 import org.elasticsearch.common.logging.ESLogger;
 import org.elasticsearch.common.logging.Loggers;
-import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.concurrent.EsRejectedExecutionException;
+import org.elasticsearch.threadpool.ThreadPool;
 
+import java.util.ArrayDeque;
+import java.util.Iterator;
 import java.util.Locale;
-import java.util.concurrent.Executors;
-import java.util.concurrent.ScheduledExecutorService;
-import java.util.concurrent.Semaphore;
-import java.util.concurrent.TimeUnit;
+import java.util.Queue;
+import java.util.concurrent.*;
 import java.util.concurrent.atomic.AtomicInteger;
 
-import static org.elasticsearch.common.util.concurrent.EsExecutors.daemonThreadFactory;
 
 /**
  * coordinates bulk operation retries for one node
  */
 public class BulkRetryCoordinator {
 
     private static final ESLogger LOGGER = Loggers.getLogger(BulkRetryCoordinator.class);
-    private static final int DELAY_INCREMENT = 1;
 
     private final ReadWriteLock retryLock;
-    private final AtomicInteger currentDelay;
+    private final BackoffPolicy backoff = LimitedExponentialBackoff.limitedExponential(1000);
 
-    private final ScheduledExecutorService retryExecutorService;
+    private final ThreadPool threadPool;
 
-    public BulkRetryCoordinator(Settings settings) {
-        this.retryExecutorService = Executors.newSingleThreadScheduledExecutor(
-                daemonThreadFactory(settings, getClass().getSimpleName()));
+    private final Object pendingLock = new Object();
+    private final Queue<PendingOperation<ShardRequest, ShardResponse>> pendingOperations = new ArrayDeque<>();
+    private volatile int activeOperations = 0;
+
+    public BulkRetryCoordinator(ThreadPool threadPool) {
+        this.threadPool = threadPool;
         this.retryLock = new ReadWriteLock();
-        this.currentDelay = new AtomicInteger(0);
     }
 
-    public ReadWriteLock retryLock() {
-        return retryLock;
+    private void trace(String message, Object ... args) {
+        if (LOGGER.isTraceEnabled()) {
+            LOGGER.trace("BulkRetryCoordinator{activeOperations='" + activeOperations + "', " +
+                         "pendingOperations='" + pendingOperations.size() + "'} {}",
+                String.format(Locale.ENGLISH, message, args));
+        }
+    }
+
+    public void acquireReadLock() throws InterruptedException {
+        retryLock.acquireReadLock();
```

Comment:
imho this wrapper method is not needed and it would miss the related releaseReadLock method. pls remove